### PR TITLE
Janusz gfal2

### PR DESCRIPTION
**Pull request by martynia**
Revised version of gfal2 top level scripts from workers.

### CODE REVIEW

File: src/pdm/workqueue/scripts/pdm_gfal2_copy.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    gfal2.cred_set(ctx, s_root, s_cred)
+    gfal2.cred_set(ctx, d_root, t_cred)
+
+    # result = []
+    for file_pair in copy_list:
+        try:
+            res = ctx.filecopy(params, str(file_pair[0]), str(file_pair[1]))
+            # result.append({'Code': res, 'Reason': 'OK'})
+            print json.dumps({'Code': res, 'Reason': 'OK'})
```
Comment (sfayer) [inline src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: Hmm, you should probably use sys.stdout.write... print may have its own buffers.

Comment (sfayer) [reply src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: Or even better, json.dump(sys.stdout, object)

Comment (martynia) [reply src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: Are you telling me that print buffer will not be flashed ? (line 100)
sys.stdout.flush()
I also noticed a bug in the monitor, when I'm flashing an incorrect stream. Fixed that. (copy job)

Comment (alexanderrichards) [reply src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: I think `print` will append a newline character as well. Better to do as @sfayer suggests
```python
json.dump(sys.stdout, object)
```

File: src/pdm/workqueue/scripts/pdm_gfal2_copy.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    if _logger.isEnabledFor(logging.DEBUG):
+        for f_source, f_dest in copy_list:
+            _logger.debug("gfal copy source: %s TO dest: %s , overwrite ? %s ",
+                          f_source, f_dest, overwrite)
+
+    s_cred = _get_cred(s_cred_file)
+    t_cred = _get_cred(t_cred_file)
+
+    if s_cred is None or t_cred is None:
+        _logger.fatal("FATAL: Please provide credential location: source %s, dest %s. ",
```
Comment (alexanderrichards) [inline src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: fatal isn't a python logging level by default, did you add this? perhaps you mean critical?

File: src/pdm/workqueue/scripts/pdm_gfal2_copy.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                      s_cred, t_cred)
+        print json.dumps({"Reason": "No credentials passed in", "Code": 1})
```
Comment (alexanderrichards) [inline src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: see @sfayer other comment

File: src/pdm/workqueue/scripts/pdm_gfal2_copy.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+    # result = []
+    for file_pair in copy_list:
+        try:
+            res = ctx.filecopy(params, str(file_pair[0]), str(file_pair[1]))
+            # result.append({'Code': res, 'Reason': 'OK'})
+            print json.dumps({'Code': res, 'Reason': 'OK'})
+            sys.stdout.flush()
+        except gfal2.GError as gerror:
+            print json.dumps({'Code': 1, 'Reason': str(gerror)})
```
Comment (alexanderrichards) [inline src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: ditto

File: src/pdm/workqueue/scripts/pdm_gfal2_ls.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    max_depth = max(-1, max_depth)
+
+    ctx = gfal2.creat_context()
+    result = OrderedDict()
+    # determine if the path point to a file, no recursion if True
+    try:
+        a = ctx.stat(root)
+    except Exception as e:
+        _logger.error("Error when obtaining ctx.stat(%s) \n %s", root, e)
+        print {'Reason': str(e), 'Code': 1}
```
Comment (alexanderrichards) [inline src/pdm/workqueue/scripts/pdm_gfal2_ls.py]: should this be json dumped?

File: src/pdm/workqueue/scripts/pdm_gfal2_ls.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    :return: None
+    """
+
+    _logger.debug("gfal listing root: %s at depth: %d", root, depth)
+
+    try:
+        a = ((item, ctx.stat(os.path.join(root, item))) for item in ctx.listdir(root))
+    except Exception as e:
+        _logger.error("Error when analysing %s \n %s", root, e)
+        print {'Reason': str(e), 'Code': 1}
```
Comment (alexanderrichards) [inline src/pdm/workqueue/scripts/pdm_gfal2_ls.py]: json dumped?

File: src/pdm/workqueue/scripts/pdm_gfal2_ls.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                break
+            dir_entry = {k: getattr(stats, k) for k, _ in
+                         inspect.getmembers(stats.__class__, lambda x: isinstance(x, property))}
+            dir_entry['name'] = dirent.d_name
+            dir_entries.append(dir_entry)
+
+        result[root] = dir_entries
+    except Exception as e:
+        _logger.error("Error when analysing %s \n %s", root, e)
+        print {'Reason': str(e), 'Code': 1}
```
Comment (alexanderrichards) [inline src/pdm/workqueue/scripts/pdm_gfal2_ls.py]: json dumped?

File: src/pdm/workqueue/scripts/pdm_gfal2_copy.py (hunk 4 above)
Comment (alexanderrichards) [reply src/pdm/workqueue/scripts/pdm_gfal2_copy.py]: there are more examples of this but I didnt flag them all

**Comment by alexanderrichards** [comment]
> So it works as advertised.

hmm ok, it's not documented as a level in python logging module so I'm surprised it works

**Comment by martynia** [comment]
log4j is guilty here:

https://stackoverflow.com/questions/31170317/what-is-the-difference-between-logging-fatal-and-logging-critical?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa <https://stackoverflow.com/questions/31170317/what-is-the-difference-between-logging-fatal-and-logging-critical?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa>

J

**Comment by alexanderrichards** [comment]
```python
>>> import logging
>>> logging.basicConfig()
>>> logging.getLogger().fatal("test")
CRITICAL:root:test
```

looks like it's an alias for critical. I wonder why it records FATAL in your example.

**Comment by martynia** [comment]
Simply, because I print FATAL literally ;-)

JM

**Comment by alexanderrichards** [comment]
ah yes! I forgot that!

**Comment by martynia** [comment]
BTW,

Is the thing not another way around ?

https://docs.python.org/2/library/json.html <https://docs.python.org/2/library/json.html>

>>> json.dump({'Code': 0, 'Reason': 'OK'}, sys.stdout)
{"Reason": "OK", "Code": 0}>>> 

(are you happy w/o the CR ?

Serialising stdout does not work well ;-) :

>>> json.dump(sys.stdout, {'Code': 0, 'Reason': 'OK'})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/json/__init__.py", line 189, in dump
    for chunk in iterable:
  File "/usr/lib64/python2.7/json/encoder.py", line 442, in _iterencode
    o = _default(o)
  File "/usr/lib64/python2.7/json/encoder.py", line 184, in default
    raise TypeError(repr(o) + " is not JSON serializable”)

JM

**Comment by alexanderrichards** [comment]
> Is the thing not another way around ?

yes, should be:
```python
json.dump(object, sys.stdout)
```
as you point out, serialising the stdout doesn't work well :wink:
> (are you happy w/o the CR ?

yes fine by me

**Comment by martynia** [comment]
I still need to test the copy code. listing and rm works fine.

**Pull request merged**